src/group_theory/monoid_localization.lean

```diff
@@ -406,8 +439,354 @@ lemma map_map {A} [comm_monoid A] {V} {g : Z →* A}
+⟨λ x, completion.mk x 1, rfl, λ x y,
+  show _ = completion.mk (x * y) (1 * 1), by rw mul_one⟩
+
+variables {X}
+
+@[simp, to_additive] lemma mk_mul_cancel_right (x y) :
+  completion.mk (x * y) y = completion.of X x :=
+by rw [←one_mul y, ←mk_mul_mk]; simpa
+
+@[simp, to_additive] lemma mk_mul_cancel_left (x : X) (y : Y) :
+  completion.mk ((y : X) * x) y = completion.of X x :=
+by rw [mul_comm, mk_mul_cancel_right]
+
+@[to_additive] instance : has_inv (completion X) :=
+⟨λ y, con.lift_on y (λ (x : X × X), completion.mk x.2 x.1) $
+  λ a b h, let ⟨c, hc⟩ := completion.eq'.1 $ completion.eq.2 h in
+  completion.eq'.2 ⟨c, by rw [mul_comm b.2, hc, mul_comm a.2]⟩⟩
+
+@[simp, to_additive] lemma inv_apply (x :  X × X) :
+(completion.mk x.1 x.2)⁻¹ = completion.mk x.2 x.1 := rfl
+
+@[simp, to_additive] protected lemma mul_left_inv (x : completion X) : x⁻¹ * x = 1 :=
+begin
+  apply con.induction_on x,
+  intro y,
+  show completion.mk _ _ = 1,
+  rw [mul_comm, mk_self],
+end
+
+@[to_additive] instance : comm_group (completion X) :=
+{  inv := has_inv.inv,
+   mul_left_inv := completion.mul_left_inv,
+   ..con.comm_monoid _ }
+
+@[to_additive] lemma mk_eq (x y) :
+  completion.mk x y = completion.of X x * (completion.of X y)⁻¹ :=
+eq_mul_inv_of_mul_eq $ by erw mk_mul_mk; rw [mul_comm, ←mk_mul_mk]; simpa
+
+variables (X)
+
+/-- The Grothendieck group of a `comm_monoid` `X` is (isomorphic to) the localization of `X` at
+    itself. This is an isomorphism rather than an equality, because the Grothendieck group is
+    defined as a quotient of `X × X`, rather than of `X × (⊤ : submonoid X)`, for clarity. -/
+@[to_additive "The Grothendieck group of an `add_comm_monoid` `X` is isomorphic to the localization of `X` at itself. This is an isomorphism rather than an equality, because the Grothendieck group is defined as a quotient of `X × X`, rather than of `X × (⊤ : submonoid X)`, for clarity."]
+def completion_equiv_localization :
+  completion X ≃* monoid_localization X (⊤ : submonoid X) :=
+{ to_fun := λ y, con.lift_on y (λ x, monoid_localization.mk x.1
+    (⟨x.2, trivial⟩ : (⊤ : submonoid X))) $ λ _ _ h, (con.eq _).2 h,
+  inv_fun := λ y, con.lift_on y (λ x, completion.mk x.1 x.2) $
+    λ a b h, (con.eq _).2 $ by rw [←@prod.mk.eta _ _ a, ←@prod.mk.eta _ _ b,
+      ←subtype.eta a.2 trivial, ←subtype.eta b.2 trivial] at h; exact h,
+  left_inv := λ y, completion.induction_on y $ λ _ _, rfl,
+  right_inv := λ y, monoid_localization.induction_on y $ λ x,
+    begin
+      dsimp,
+      erw monoid_localization.lift_on_beta,
+        dsimp,
+        rw subtype.coe_eta,
+        refl,
+      exact λ _ _, (con.eq _).2,
+    end,
+  map_mul' := λ x y, con.induction_on₂ x y $ λ _ _, rfl }
+
+variables {X}
+
+/-- The localization of a `comm_group` `G` at itself is (isomorphic to) `G`. -/
+@[to_additive "The localization of an `add_comm_group` `G` at itself is (isomorphic to) `G`."]
+def mul_equiv_of_group (G : Type*) [comm_group G] :
```

Comment:
Ideally, this would follow in a couple of lines from the `char_pred` machinery.

Reply:
As could ```completion_equiv_localization``` and ```mul_equiv_map```, but since you can give explicit inverses for these I thought I might as well. I will say I have no idea why constructive definitions are preferable. I just instinctively avoid using ```classical.some, classical.some_spec``` due to a long time of getting confused by them before I started using ```set_option pp.proofs true``` 😄